### PR TITLE
bpo-40636: Clarify the zip built-in docstring.

### PR DESCRIPTION
**Pull request by gpshead**
This puts much simpler text up front along with an example.

As it was, the zip built-in docstring was technically correct.  But too
technical for the reader who shouldn't _need_ to know about `__next__` and
`StopIteration` as most people do not need to understand the internal
implementation details of the iterator protocol in their daily life.

This is a documentation only change, intended to be backported to 3.8; it is
only tangentially related to PEP-618 which might offer new behavior options
in the future.

https://bugs.python.org/issue40636

### CODE REVIEW

File: Python/bltinmodule.c [hunk 1]
```diff
@@ -2649,9 +2649,14 @@ static PyMethodDef zip_methods[] = {
 };
 
 PyDoc_STRVAR(zip_doc,
-"zip(*iterables) --> zip object\n\
+"zip(*iterables) --> An iterable zip object that yields tuples.\n\
```
Comment (gpshead) [inline Python/bltinmodule.c]: To make the single line help text more useful, I'm pondering changing this to

`A zip object that yields tuples until one input is exhausted.`

**Comment by brandtbucher** [comment]
I really like the docstring for `enumerate`. I wonder if we could follow its example:

```
Return an enumerate object.

  iterable
    an object supporting iteration

The enumerate object yields pairs containing a count (from start, which
defaults to zero) and a value yielded by the iterable argument.

enumerate is useful for obtaining an indexed list:
    (0, seq[0]), (1, seq[1]), (2, seq[2]), ...
```

**Comment by gpshead** [comment]
Good suggestion, take a look at the latest proposed wording.

File: Python/bltinmodule.c [hunk 2]
```diff
@@ -2649,12 +2649,17 @@ static PyMethodDef zip_methods[] = {
...
-is exhausted and then it raises StopIteration.");
+Upon any of the input iterables being exhausted, the zipping stops.\n\
+\n\
+   >>> list(zip('abcdefg', range(3), range(4)))\n\
+   [('a', 0, 0), ('b', 1, 1), ('c', 2, 2)]\n\
+\n\
+The zip object yields n-length tuples where n is the number of iterables\n\
+passed as positional arguments to zip().  The i-th element in every tuple\n\
+comes from the i-th iterable argument to zip().  This continues until the\n\
+shortest iterable in the argument sequences is exhausted.");
```
Comment (gpshead) [inline Python/bltinmodule.c]: i should probably replace "sequences" with "iterables" here.

File: Python/bltinmodule.c [hunk 3]
```diff
@@ -2649,12 +2649,17 @@ static PyMethodDef zip_methods[] = {
...
 PyDoc_STRVAR(zip_doc,
-"zip(*iterables) --> zip object\n\
+"zip(*iterables) --> A zip object yielding tuples until an input is exhausted.\n\
 \n\
-Return a zip object whose .__next__() method returns a tuple where\n\
-the i-th element comes from the i-th iterable argument.  The .__next__()\n\
-method continues until the shortest iterable in the argument sequence\n\
-is exhausted and then it raises StopIteration.");
+Upon any of the input iterables being exhausted, the zipping stops.\n\
+\n\
```
Comment (brandtbucher) [inline Python/bltinmodule.c]: I'd strike this line. We already repeat this warning twice elsewhere.

```suggestion
```

File: Python/bltinmodule.c [hunk 4]
```diff
@@ -2649,12 +2649,17 @@ static PyMethodDef zip_methods[] = {
...
+   >>> list(zip('abcdefg', range(3), range(4)))\n\
+   [('a', 0, 0), ('b', 1, 1), ('c', 2, 2)]\n\
+\n\
+The zip object yields n-length tuples where n is the number of iterables\n\
```
Comment (brandtbucher) [inline Python/bltinmodule.c]:
```suggestion
The zip object yields n-length tuples, where n is the number of iterables\n\
```

File: Python/bltinmodule.c [hunk 5]
```diff
@@ -2649,12 +2649,17 @@ static PyMethodDef zip_methods[] = {
...
+passed as positional arguments to zip().  The i-th element in every tuple\n\
+comes from the i-th iterable argument to zip().  This continues until the\n\
+shortest iterable in the argument iterables is exhausted.");
```
Comment (brandtbucher) [inline Python/bltinmodule.c]: A bit wordy... we use the word "iterable" a lot already:

```suggestion
shortest argument is exhausted.");
```

**Review by brandtbucher — CHANGES_REQUESTED** [review]
Thanks for this... just a couple of suggestions.

**Comment by gpshead** [comment]
good suggestios, this is looking good.

**Review by brandtbucher — APPROVED** [review]
